src/concurrent.cr

```diff
@@ -128,14 +128,110 @@ macro spawn(call, *, name = nil)
   {% end %}
 end
 
+# Runs the commands passed as arguments concurrently (in Fibers) and waits
+# for them to finish.
+#
+# ```
+# def say(word)
+#   puts word
+# end
+#
+# # Will print out the three words concurrently
+# parallel(
+#   say("concurrency"),
+#   say("is"),
+#   say("easy")
+# )
+# ```
+#
+# Can also be used to conveniently collect the return values of the
+# concurrent operations.
+#
+# ```
+# def concurrent_job(word)
+#   word
+# end
+#
+# a, b, c =
+#   parallel(
+#     concurrent_job("concurrency"),
+#     concurrent_job("is"),
+#     concurrent_job("easy")
+#   )
+#
+# puts a # => "concurrency"
+# puts b # => "is"
+# puts c # => "easy"
+# ```
+#
+# Exceptions raised within the concurrent operations will be re-raised
+# in the parent thread. If you'd like to avoid this, consider using the
+# `parallel?` macro instead.
 macro parallel(*jobs)
-  %channel = Channel(Nil).new
+  %channel = Channel(Exception | Nil).new
 
   {% for job, i in jobs %}
     %ret{i} = uninitialized typeof({{job}})
     spawn do
       begin
         %ret{i} = {{job}}
+      rescue e : Exception
+        %channel.send e
+      else
+        %channel.send nil
+      end
+    end
+  {% end %}
+
+  {{ jobs.size }}.times do
+    %value = %channel.receive
+    if %value.is_a?(Exception)
+      raise %value
```

Comment:
The problem here is that this exception doesn't contain the backtrace for the location of the parallel statement, it only contains the backtrace for the new fiber stack. Other languages wrap exceptions with a "caused by" to solve this, then you get a backtrace to the location of the `parallel` call and another backtrace of the error inside the new fiber.

Crystal has a `cause` field in `Exception`, but it doesn't print it in `inspect_with_backtrace`.

Reply:
do we want to add the cause exception to `inspect_with_backtrace`? (that's probably not in the scope of this PR though)

if so, then I could raise a new exception with the `cause` pointing to the original one. for the moment it probably wouldn't help much but when/if the backtrace starts containing the cause, this point will at least be covered.

in my opinion, pointing to the source of the error (the fiber) is more important than the re-raising point, though the latter might help in some situations as well.

Reply:
I guess we should improve `cause` then?